01-calculator.c

```diff
@@ -0,0 +1,143 @@
+int precedence(char op)
+{
+    if (op == '+' || op == '-')
+        return 1;
+    return 2;
+}
+
+int applyOperation(int a, int b, char op)
+{
+    if (op == '+')
+        return a + b;
+    if (op == '-')
+        return a - b;
+    if (op == '*')
+        return a * b;
+    if (op == '/')
+    {
+        if (b == 0)
+            Error("Error:Division by zero");
+        return a / b;
+    }
+}
+char operatorsAllowed[] = "/*+-";
+
+// This function check for invalid input like alphabest,special characters and return true if found
+bool invalidInput(char inp)
+{
+    if (strchr(operatorsAllowed, inp) == NULL && !isdigit(inp))
+        return true;
+    return false;
+}
+
+// This function pop two operators and operand and evaluate
+void evaluate(Stack *operands, Stack *operators)
+{
+    int b = pop(operands);
+    int a = pop(operands);
+    char op = pop(operators);
+    push(operands, applyOperation(a, b, op));
+}
+
+int evaluateExpression(const char *expression)
+{
+    Stack operators, operands;
+    operands.topIndex = operators.topIndex = -1;
+    // expression always starts with an operand.
+    bool operandTurn = true;
+    int i = 0, n = strlen(expression);
+    while (i < n)
+    {
+        if (isspace(expression[i]))
+            i++;
+        else if (invalidInput(expression[i]))
+            Error("Error:Invalid Input");
+        // If the input is a number,parse it to integer
+        // if input->+3+4 and operandTurn is true but first input is operator, means input is invalid
+        else if (isdigit(expression[i]) && operandTurn)
+        {
+            int value = 0;
+            while (i < n && isdigit(expression[i]))
+            {
+                value = value * 10 + (expression[i] - '0');
+                i++;
+            }
+            push(&operands, value);
+            operandTurn = false;
+        }
+        // if the input is an operator
+        else if (strchr(operatorsAllowed, expression[i]) != NULL && !operandTurn)
+        {
+            // if input->12++45 then after examine 12+ operantTurn=false and again when found + but operandTurn is true,means expression is invalid
+
+            //  While top of operators has the same or greater precedence
+            while (operators.topIndex != -1 && precedence(top(&operators)) >= precedence(expression[i]))
+                evaluate(&operands, &operators);
+            push(&operators, expression[i]);
+            operandTurn = true;
+            i++;
+        }
+        else
+            Error("Error:Invalid Input");
+    }
+    // if the expression is valid , we always end up with operandTurn=false
+    //  if the input->3+4+ then operandTurn is true,means expression is invalid
+    if (operandTurn)
+        Error("Error:Invalid Input");
+
+    // Process remaining expression
+    while (operators.topIndex != -1)
+        evaluate(&operands, &operators);
+
+    // Final result
+    return pop(&operands);
+}
+
+int main()
```

Comment:
In main() function, there should be only function calling.
There should not be things like "taking inputs", etc.